File: 1035-cousins-in-binary-tree/cousins-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        queue<pair<TreeNode*,TreeNode*>>q;
        q.push({root,nullptr});
        while(!q.empty()){
            int size=q.size();
            TreeNode* p_x=nullptr;
            TreeNode* p_y=nullptr;
            for(int i=0;i<size;i++){
                pair<TreeNode*,TreeNode*> curr=q.front();
                q.pop();
                if(curr.first->val==x){
                    p_x=curr.second;
                }
                if(curr.first->val==y){
                    p_y=curr.second;
                }
                if(curr.first->left)q.push({curr.first->left,curr.first});
                if(curr.first->right)q.push({curr.first->right,curr.first});
            }
            if(p_x!=nullptr && p_y!=nullptr && p_x!=p_y)return true;
        }
        return false;

        
    }
};

File: 1035-cousins-in-binary-tree/cousins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Pair {
    TreeNode curr;
    TreeNode parent;
    Pair(TreeNode curr,TreeNode parent){
        this.curr=curr;
        this.parent=parent;
    }
}
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        Queue<Pair>q = new LinkedList<>();
        if(root==null){
            return false;
        }
        q.add(new Pair(root,null));
        while(!q.isEmpty()){
            int size=q.size();
            TreeNode px = null;
            TreeNode py = null;
            for(int i=0;i<size;i++){
                Pair p = q.poll();
                TreeNode node = p.curr;
                TreeNode node_parent = p.parent;
                if(node.val==x){
                    px = node_parent;
                }
                if(node.val==y){
                    py = node_parent;
                }
                if(node.left!=null){
                    q.add(new Pair(node.left,node));
                }
                if(node.right!=null){
                    q.add(new Pair(node.right,node));
                }
            }
            if(px!=null && py!=null && px!=py){
                return true;
            }
        }
        return false;
    }
}